registry/reference.go

```diff
@@ -159,20 +180,36 @@ func (r Reference) ValidateRepository() error {
 	return nil
 }
 
-// ValidateReference validates the reference.
-func (r Reference) ValidateReference() error {
-	if r.Reference == "" {
-		return nil
-	}
-	if _, err := r.Digest(); err == nil {
-		return nil
-	}
+// ValidateReferenceAsTag validates the reference as a tag.
+func (r Reference) ValidateReferenceAsTag() error {
 	if !tagRegexp.MatchString(r.Reference) {
 		return fmt.Errorf("%w: invalid tag", errdef.ErrInvalidReference)
 	}
 	return nil
 }
 
+// ValidateReferenceAsDigest validates the reference as a digest.
+func (r Reference) ValidateReferenceAsDigest() error {
+	if _, err := r.Digest(); err != nil {
+		return fmt.Errorf("%w: invalid digest; %v", errdef.ErrInvalidReference, err)
+	}
+	return nil
+}
+
+// ValidateReference where the reference is first tried as an ampty string, then
+// as a digest, and if that fails, as a tag.
+func (r Reference) ValidateReference() error {
+	if len(r.Reference) == 0 {
+		return nil
+	}
+
+	if err := r.ValidateReferenceAsDigest(); err != nil {
+		return err
+	}
+
+	return r.ValidateReferenceAsTag()
```

Comment:
Inspired by your code change in `repository.go`, we can have a smarter implementation of `ValidateReference()`.

```suggestion
	if index = strings.IndexByte(reference, ':'); index != -1 {
		return r.ValidateReferenceAsDigest()
	}
	return r.ValidateReferenceAsTag()
```

Reply:
Nice, fixed :)